File: February challenges/Java/76. Minimum Window Substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (t.equals("")) return "";

        Map<Character, Integer> countT = new HashMap<>();
        Map<Character, Integer> windowT = new HashMap<>();

        for (char i : t.toCharArray()) {
            countT.put(i, 1 + countT.getOrDefault(i, 0));
        }

        int have = 0;
        int need = countT.size();
        int resLen = Integer.MAX_VALUE;
        int[] res = new int[]{-1, -1};
        int l = 0;

        for (int r = 0; r < s.length(); r++) {
            char c = s.charAt(r);
            windowT.put(c, 1 + windowT.getOrDefault(c, 0));

            if (countT.containsKey(c) && windowT.get(c).equals(countT.get(c))) {
                have++;
            }

            while (have == need) {
                if ((r - l + 1) < resLen) {
                    res[0] = l;
                    res[1] = r;
                    resLen = (r - l + 1);
                }

                windowT.put(s.charAt(l), windowT.get(s.charAt(l)) - 1);
                if (countT.containsKey(s.charAt(l)) && windowT.get(s.charAt(l)) < countT.get(s.charAt(l))) {
                    have--;
                }

                l++;
            }
        }

        int left = res[0];
        int right = res[1];
        return left != -1 ? s.substring(left, right + 1) : "";
    }
}

File: February challenges/Python/76 Minimum Window Substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == "": return ""

        countT, windowT = {}, {}

        for i in t:
            countT[i] = 1 + countT.get(i, 0)

        have, need = 0, len(countT)
        resLen = float("infinity")
        res = [-1, -1]
        l = 0
        for r in range(len(s)):
            c = s[r]
            windowT[c] = 1 + windowT.get(c, 0)

            if c in countT and windowT[c] == countT[c]:
                have += 1

            while have == need:
                if (r - l + 1) < resLen:
                    res = [l, r]
                    resLen = (r - l + 1)

                windowT[s[l]] -= 1
                if s[l] in countT and windowT[s[l]] < countT[s[l]]:
                    have -= 1
                
                l += 1

        l, r = res
        return s[l:r+1] if resLen != float("infinity") else ""
 

            
